### PR TITLE
Technical advisory for race condition between long running reads and …

### PR DESCRIPTION
**Pull request by lunevalex**
…replica removal.

### CODE REVIEW

**Comment by nickelnickel** [comment]
Small tweaks to first paragraph in **Description**:

Cockroach Labs has discovered a race condition, where a long running read request submitted during replica removal may be evaluated on the removed replica, returning an empty result. Common causes of replica removal are range merges and replica rebalancing in the cluster. When a replica is removed there is a small window, where the data is already deleted but a read request will still see it as a valid replica. In this case a read request will always return an empty result for the part of the query that relied on this range.

-----

And to **Impact**:

This issue may impact all read operations, while the window of opportunity to see is very limited it may have happened to without your knowledge. As read operations include backup and incremental backup, if your cluster has experienced overloaded conditions we encourage you to verify your backups using the instructions provided above.

----

Open question as discussed on Slack about: `- The replica is marked for removal and data is cleared.`

does the read have to start before and finish after all the liveness, lease change, merge operations have completed?

File: advisories/a64325.md
Comment (nvanbenschoten) [inline advisories/a64325.md:18]: Consider re-wording to "the old leaseholder replica from the range".

Also, consider adding "The new leaseholder _decides to_" to hint that this is another independent condition that needs to hold.

File: advisories/a64325.md
Comment (nvanbenschoten) [inline advisories/a64325.md:31]: I wouldn't link to #46329. While it's related, it's not the same issue. In some ways, it's worse, because we weren't even trying to get the synchronization right. But it's also much less severe in its potential outcomes and we have no evidence that anyone has actually run into it.

**Review by nvanbenschoten — APPROVED** [review]

**Comment by lunevalex** [comment]
@nickelnickel updated

**Comment by nickelnickel** [comment]
LGTM pending the addition from @shermanCRL .

**Review by mwang1026 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10478)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lunevalex, @nickelnickel, and @shermanCRL)

___
*[advisories/a64325.md, line 29 at r3](https://reviewable.io/reviews/cockroachdb/docs/10478#-MZZ_aSZ9ueVBZQQEft8:-MZZ_aSZ9ueVBZQQEft9:b-2261sr) ([raw file](https://github.com/cockroachdb/docs/blob/c72a32f7a36add73d3cab56d431ccf7cb8b75e54/advisories/a64325.md#L29)):*
> ```Markdown
> This is resolved in CockroachDB by PR [#64324], which fix the race condition between read requests and replica removal.
> 
> The fix has been applied to maintenance releases of CockroachDB v20.1, v20.2, and v21.1.
> ```

Have we considered mentioning the exact release? I feel like people would ask "what do I need to upgrade to"?

File: advisories/a64325.md
Comment (shermanCRL) [inline advisories/a64325.md:13]: I’d rephrase as:

This is a rare scenario that can only be encountered if a number of conditions are met:

It’s better not to speculate on likelihood. State the facts and let operators determine likelihood. I do think “rare” is a safe characterization, but we might even consider removing that.

File: advisories/a64325.md
Comment (shermanCRL) [inline advisories/a64325.md:29]: Are these released? We should say what patch version. It’d be convenient to offer the releases page link here too.

File: advisories/a64325.md
Comment (shermanCRL) [inline advisories/a64325.md:37]: @pbardea Would you fill this in? I assume it’s a summary of this: https://github.com/cockroachlabs/support/issues/940#issuecomment-828814334

It’s important to say what the query can or can’t detect.

--

Format-wise, because these instructions are long, maybe they don’t belong in this document.

Should they be added as a comment on the public issue? https://github.com/cockroachdb/cockroach/issues/64325

Thoughts @lunevalex ?

Comment (pbardea) [reply advisories/a64325.md:37]: Will follow up. There are a few queries that can help detect these missing from full backups. I've posted them in https://github.com/cockroachlabs/support/issues/940#issuecomment-830420092. They each have tradeoffs in terms of how many edge cases they consider and the number of false positives returned. They're all fairly clunky, especially to explain in general terms in a TA.

Perhaps the best solution here is to find a way to detect these issues with a high-probability but with minimal false-positives reported?

Of note: I don't think this issue is detectable at all in incremental backups.

Comment (shermanCRL) [reply advisories/a64325.md:37]: It might be a better decision not to include this in the advisory for now, too much detail. We’ll use the above comment as a resource for the support team, and think about where a public version would make sense.

Comment (shermanCRL) [inline advisories/a64325.md:37]: Maybe a rephrasing of this paragraph:

Because this issue affects reads, it may impact backups. Our [support team](https://support.cockroachlabs.com/) has tools to detect and troubleshoot where this might be true, please contact them.

File: advisories/a64325.md
Comment (erikgrinaker) [inline advisories/a64325.md:24]: Backported a test for this to 19.2 (#64510), and as far as I can tell it does not have this bug -- it appears to have been introduced in 20.1.

File: _data/advisories.yml
Comment (erikgrinaker) [inline _data/advisories.yml:2]:
```suggestion
  summary: Race condition between reads and replica removal
```

File: _data/advisories.yml
Comment (erikgrinaker) [inline _data/advisories.yml:3]:
```suggestion
  versions: 20.1 and later
```

File: advisories/a64325.md
Comment (erikgrinaker) [inline advisories/a64325.md:27]:
```suggestion
This is resolved in CockroachDB by PR [#64324], which fixes the race condition between read requests and replica removal.
```

File: advisories/a64325.md
Comment (erikgrinaker) [inline advisories/a64325.md:18]:
```suggestion
 - The new leaseholder decides to immediately remove the old leaseholder replica from the range, due to a rebalance or range merge.
```

**Review by erikgrinaker — COMMENTED** [review]
Note that read/write requests have a similar bug that hasn't been fixed yet, PR #64471 has been submitted for this.

File: advisories/a64325.md
Comment (lunevalex) [reply advisories/a64325.md:24]: Thanks for doing that

File: advisories/a64325.md
Comment (lunevalex) [reply advisories/a64325.md:29]: Not yet, we will request extraordinary releases for these next week

File: advisories/a64325.md
Comment (lunevalex) [reply advisories/a64325.md:37]: I think we can go with a link and keep the advisory simple.

**Comment by lunevalex** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10478)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lunevalex, @nickelnickel, and @shermanCRL)
> 
> _[advisories/a64325.md, line 29 at r3](https://reviewable.io/reviews/cockroachdb/docs/10478#-MZZ_aSZ9ueVBZQQEft8:-MZZ_aSZ9ueVBZQQEft9:b-2261sr) ([raw file](https://github.com/cockroachdb/docs/blob/c72a32f7a36add73d3cab56d431ccf7cb8b75e54/advisories/a64325.md#L29)):_
> 
> > ```gfm
> > This is resolved in CockroachDB by PR [#64324], which fix the race condition between read requests and replica removal.
> > 
> > The fix has been applied to maintenance releases of CockroachDB v20.1, v20.2, and v21.1.
> > ```
> 
> Have we considered mentioning the exact release? I feel like people would ask "what do I need to upgrade to"?

It's mentioned in the mitigation section

File: advisories/a64325.md
Comment (shermanCRL) [inline advisories/a64325.md:39]: LGTM

**Review by nvanbenschoten — APPROVED** [review]

File: advisories/a64325.md
Comment (erikgrinaker) [inline advisories/a64325.md:11]:
```suggestion
Cockroach Labs has discovered a race condition, where a read request submitted during replica removal may be evaluated on the removed replica, returning an empty result. Common causes of replica removal are range merges and replica rebalancing in the cluster. When a replica is removed there is a small window, where the data is already deleted but a read request will still see it as a valid replica. In this case a read request will always return an empty result for the part of the query that relied on this range.
```

File: advisories/a64325.md
Comment (erikgrinaker) [inline advisories/a64325.md:35]:
```suggestion
Users of CockroachDB v20.1 or v20.2 are invited to upgrade to v20.1.16, v20.2.9, or a later version.
```

File: advisories/a64325.md
Comment (erikgrinaker) [inline advisories/a64325.md:39]:
```suggestion
This issue may impact all read operations. While the window of opportunity to see it is very limited, it may have happened without your knowledge. Because this issue affects reads, it may impact backups. Our [support team](https://support.cockroachlabs.com/) has tools to detect and troubleshoot where this might be true, please contact them.
```

**Review by erikgrinaker — APPROVED** [review]

**Review by pbardea — APPROVED** [review]

**Review by nickelnickel — APPROVED** [review]
This looks good. My only suggestions are minor:

- Users of CockroachDB v20.1 or v20.2 are invited to upgrade to v20.1.16, v20.2.9, or a later version.
+ Users of CockroachDB v20.1 or v20.2 are encouraged to upgrade to v20.1.16, v20.2.9, or a later version.

- Our [support team](https://support.cockroachlabs.com/) has tools to detect and troubleshoot where this might be true, please contact them.
- Our [support team](https://support.cockroachlabs.com/) has tools to detect and troubleshoot this condition. If you would like help to check this in your environment, please contact them.

**Pull request merged**
